Add collisionExtendsVertically

yrsegal opened the pull request:
Example implementation: https://gist.github.com/yrsegal/927e018abb33505fda2d98c05907ce3f

In prior versions (1.12-), entities checked the blockspace below them unconditionally for collision with hitboxes that extend above the block space.
In 1.13, this was changed by Mojang to only collide when the block below is a fence, fence gate, or wall. This breaks the collision of any modded block whose collision box extends above its block space.
This PR allows blocks to opt back in to that behaviour.

yrsegal committed:
Add collisionExtendsVertically

This allows non-fence non-wall blocks to collide above their hitboxes.

ichttt commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:
```diff
@@ -1024,4 +1024,13 @@ default void onBlockExploded(BlockState state, World world, BlockPos pos, Explos
+     */
+    default boolean collisionExtendsVertically(BlockState state, IBlockReader world, BlockPos pos, Entity collidingEntity)
+    {
+        return false;
```
I'd argue we should change it to use the two contains checks on the tags that vanilla always did, and move remove the check from the original method, so you can override this to false even if it is in the fences or wall tags. Just to give it even more flexibility...

yrsegal committed:
Replace default behavior entirely

The WalkNodeProcessor patch, on second analysis, is unneeded. As you can already return the node type from a block at will, there's not really a need for injecting into the "default" node type method.
